modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/LinearRankFeatureTermQuery.java

```diff
@@ -0,0 +1,304 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.mapper.extras;
+
+import org.apache.lucene.index.IndexReaderContext;
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.PostingsEnum;
+import org.apache.lucene.index.ReaderUtil;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.index.TermState;
+import org.apache.lucene.index.TermStates;
+import org.apache.lucene.index.Terms;
+import org.apache.lucene.index.TermsEnum;
+import org.apache.lucene.search.Explanation;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.LeafSimScorer;
+import org.apache.lucene.search.Matches;
+import org.apache.lucene.search.MatchesIterator;
+import org.apache.lucene.search.MatchesUtils;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.TermQuery;
+import org.apache.lucene.search.TermScorer;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.search.similarities.Similarity;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Lucene query to be used by {@link RankFeaturesFieldMapper}  when creating term queries
+ *
+ * This is adapted from Lucene's FeatureQuery query and extends TermQuery
+ */
+class LinearRankFeatureTermQuery extends TermQuery {
```

Comment:
Why are you extending `TermQuery` rather than `Query`?

Also what would be wrong with using `FeatureQuery`? I know it doesn't caching and a few other things but this would only be a minor inefficiency? As far as the limit on boosts in concerned, it's easy to work around by wrapping the query with a `BoostQuery` manually, the limit mostly exists because otherwise scores are logs so boosts higher than 64 (~ log(Long.MAX_VALUE)) would essentially ignore any other contribution from the query.

In case not returning a `TermQuery` causes problems, another option could be to use a vanilla `TermQuery` and update the similarity exposed on the `TextSearchInfo` object to implement `float SimScorer#score(float freq, long norm)` like your `decodeFeatureValue` function below.

Reply:
> Also what would be wrong with using FeatureQuery?[...] As far as the limit on boosts in concerned, it's easy to work around by wrapping the query with a BoostQuery manually

This was my main concern, but looking at how our `TermQueryBuilder` works, we do this wrapping in a `BoostQuery` when the user provides a boost in the query. Consequently, it works just fine. I have added a test to make sure that `term` queries against the field with larger boosts works.

> I know it doesn't caching and a few other things but this would only be a minor inefficiency? 

I honestly don't know what the inefficiency would be, typically this will be used in a `scoring` scenario. If we want to make it faster, I think we could attempt to update lucene.

I am comfortable with accepting any potential inefficiencies as we don't have a good data set yet to measure them :).